### PR TITLE
Adding new `choice` type

### PR DESCRIPTION
**Pull request by emulienfou**
Adding new type `choice`, this feature allow user to select a value in a ChoiceType field.

### CODE REVIEW

**Comment by emulienfou** [comment]
Hi @martiis,
I made some (unfinished) modifications to simplify everything.
However, when I save a choice value in the ORM, the `choices` data are empty in my ChoiceType.

Did I need to also save the default choices into the database ?

Here is the example code I use:
```yaml
helis_settings_manager:
  settings:
    - name: choice
      type: choice
      choices: ['choice1', 'choice2']
      data: 'choice1'

    - name: choice2
      type: choice
      choices:
        - 'choice1'
        - key: value
```

Thanks

**Comment by martiis** [comment]
> Hi @martiis,
> I made some (unfinished) modifications to simplify everything.
> However, when I save a choice value in the ORM, the `choices` data are empty in my ChoiceType.
> 
> Did I need to also save the default choices into the database ?
> 
> Here is the example code I use:
> 
> ```yaml
> helis_settings_manager:
>   settings:
>     - name: choice
>       type: choice
>       choices: ['choice1', 'choice2']
>       data: 'choice1'
> 
>     - name: choice2
>       type: choice
>       choices:
>         - 'choice1'
>         - key: value
> ```
> 
> Thanks

Now there is an ugly way of transforming model into entity. We'll improve that later on :)
https://github.com/HelisLT/settings-manager-bundle/blob/5878f54a6acd70835d67f7990f95c3426e74a69f/src/Provider/DoctrineOrmSettingsProvider.php#L204

just add choices to the entity and add choices mapping here:
https://github.com/HelisLT/settings-manager-bundle/blob/master/src/Resources/config/doctrine/Model.SettingModel.orm.yml

File: src/Settings/SettingsManager.php [hunk 1]
```diff
@@ -87,30 +87,35 @@ public function getSettingsByName(array $domainNames, array $settingNames): arra
     {
         $settings = [[]];
 
+        $settingConfigModel = $this->providers['config']->getSettingsByName($domainNames, $settingNames);
```
Comment (martiis) [inline src/Settings/SettingsManager.php]: What are these changes to SettingsManager?

File: src/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -142,6 +142,20 @@ private function getSettingsNode(): NodeDefinition
                         ->end()
                     ->end()
                 ->end()
+                ->arrayNode('choices')
+                    ->arrayPrototype()
```
[the remark this answers is not in the document]
Comment (emulienfou) [reply src/DependencyInjection/Configuration.php]: No reason to use `arrayPrototype`, I'will fix this ;)

File: src/Settings/SettingsManager.php (hunk 1 above)
Comment (emulienfou) [reply src/Settings/SettingsManager.php]: Oups! Was not suppose to push this here :(

**Comment by emulienfou** [comment]
So @martiis, I made the changes has you ask and things are way betters.

However I see an issue with the choices.
In my case, I populate the choices list of available values dynamically from a service.

By saving the choices values into the database, it cannot be dynamic.
Do you think it's something to think in this bundle, or you prefer to not handle that in this bundle, so I have to implement it on my own?

File: src/Resources/config/doctrine/Model.SettingModel.orm.yml [hunk 3]
```diff
@@ -21,6 +21,11 @@ Helis\SettingsManagerBundle\Model\SettingModel:
             name: data
             type: json_array
             nullable: false
+        choices:
+            name: choices
+            type: json_array
```
Comment (martiis) [inline src/Resources/config/doctrine/Model.SettingModel.orm.yml]: `json_array` is deprecated, use `json`
https://www.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/types.html#json-array

**Comment by martiis** [comment]
> So @martiis, I made the changes has you ask and things are way betters.
> 
> However I see an issue with the choices.
> In my case, I populate the choices list of available values dynamically from a service.
> 
> By saving the choices values into the database, it cannot be dynamic.
> Do you think it's something to think in this bundle, or you prefer to not handle that in this bundle, so I have to implement it on my own?

Had a few thoughts about your case. With current implementation we don't want a base provider from which everything else populates, they are completely separate and work in their own ways. They have their own settings which can come from anywhere. It's only SettingsManager job to know which setting to pick for you.

Having dynamic choices can be a bit slippery. What if a choice disappears when it was set to setting as its value?

A quick solution for your case could be adding new events like `settings_manager.edit_setting`, `settings_manager.edit_setting.<setting_name>` in controller, on which you could set currently available choices before putting it into form and that would work every time without needing to update anything in providers.

How that sound?

File: src/Resources/config/doctrine/Model.SettingModel.orm.yml (hunk 3 above)
Comment (emulienfou) [reply src/Resources/config/doctrine/Model.SettingModel.orm.yml]: Fixed for keys `choices` but also `data`

**Comment by emulienfou** [comment]
Hi @martiis, I will implements everything on my side.
So your solution sounds perfect for me.

**Comment by martiis** [comment]
> Hi @martiis, I will implements everything on my side.
> So your solution sounds perfect for me.

Great I'll add these events on upcoming days ( try one or two :) ) and again thanks for pr 👍

**Pull request merged**